packages/core/src/build/configAndIndexingFunctions.ts

```diff
@@ -103,56 +104,111 @@ export async function buildConfigAndIndexingFunctions({
           })
           .catch((e) => {
             throw new Error(
-              `Unable to fetch "latest" block for network '${network.name}':\n${e.message}`,
+              `Unable to fetch "latest" block for chain '${chain.chain.name}':\n${e.message}`,
             );
           });
-        perNetworkLatestBlockNumber.set(network.name, blockPromise);
+        perChainLatestBlockNumber.set(chain.chain.name, blockPromise);
         return blockPromise;
       }
     }
     return blockNumberOrTag;
   };
 
-  const networks: Network[] = await Promise.all(
-    Object.entries(config.networks).map(async ([networkName, network]) => {
-      const { chainId, transport } = network;
-
-      const defaultChain =
-        Object.values(chains).find((c) =>
-          "id" in c ? c.id === chainId : false,
-        ) ?? chains.mainnet!;
-      const chain = { ...defaultChain, name: networkName, id: chainId };
-
-      // Note: This can throw.
-      const rpcUrls = await getRpcUrlsForClient({ transport, chain });
-      rpcUrls.forEach((rpcUrl) => {
-        if (isRpcUrlPublic(rpcUrl)) {
-          logs.push({
-            level: "warn",
-            msg: `Network '${networkName}' is using a public RPC URL (${rpcUrl}). Most apps require an RPC URL with a higher rate limit.`,
-          });
+  const chains: Chain[] = await Promise.all(
+    Object.entries(config.chains).map(async ([chainName, chain]) => {
+      const matchedChain = Object.values(viemChains).find((c) =>
+        "id" in c ? c.id === chain.id : false,
+      );
+
+      if (matchedChain === undefined) {
+        throw new Error(
+          `Chain "${chainName}" with id ${chain.id} not found in viem. Please update viem to the latest version.`,
+        );
+      }
+      matchedChain.name = chainName;
+
+      if (chain.rpcUrl === undefined) {
```

Comment:
Nit but would be more readable to coalesce these three paths to an array of strings first and check each entry each once